# String Manipulation with stringr in R

Character strings can turn up in all stages of a data science project. You might have to clean messy string input before analysis, extract data that is embedded in text or automatically turn numeric results into a sentence to include in a report. Perhaps the strings themselves are the data of interest, and you need to detect and match patterns within them. This course will help you master these tasks by teaching you how to pull strings apart, put them back together and use stringr to detect, extract, match and split strings using regular expressions, a powerful way to express patterns.

## String basics
You'll start with some basics: how to enter strings in R, how to control how numbers are transformed to strings, and finally how to combine strings together to produce output that combines text and nicely formatted numbers.

### Quotes
Let's get started by entering some strings in R. In the video you saw that you use quotes to tell R 
to interpret something as a string. Both double quotes (") and single (') quotes work, but there are 
some guidelines for which to use.
First, you should prefer double quotes (") to single quotes ('). That means, whenever you are defining a 
string your first intuition should be to use ".
Unfortunately if your string has " inside it, R will interpret the double quote as "this is the end of 
the string", not as "this is the character "". This is one time you can forget the first guideline and use 
the single quote, ', to define the string.
Finally, there are cases where you need both ' and " inside the string. In this case, fall back to the 
first guideline and use " to define the string, but you'll have to escape any double quotes inside the 
string using a backslash (i.e. \").
To practice, you are going to enter a few lines from Lewis Carroll's Alice's Adventures in Wonderland. 
Alice has just arrived at the tea party…

In [2]:
# Following the guidelines for using quotes, define the three strings, line1, line2 and line3:

# Line 1: The table was a large one, but the three were all crowded together at one corner of it:
line1 <- "The table was a large one, but the three were all crowded together at one corner of it:"
# Line 2: "No room! No room!" they cried out when they saw Alice coming.
line2 <- '"No room! No room!" they cried out when they saw Alice coming.' 
# Line 3:"There's plenty of room!" said Alice indignantly, and she sat down in a large arm-chair at one end of the table.
line3 <- "\"There's plenty of room!\" said Alice indignantly, and she sat down in a large arm-chair at one end of the table."

### What you see isn't always what you have

Even though you used single quotes so you didn't have to escape any double quotes, when R prints it, 
you'll see escaped double quotes (\")! R doesn't care how you defined the string, it only knows what 
the string represents, in this case, a string with double quotes inside.
When you ask R for line2 it is actually calling print(line2) and the print() method for strings displays strings as 
you might enter them. If you want to see the string it represents you'll need to use a different function: writeLines().
You can pass writeLines() a vector of strings and it will print them to the screen, each on a new line. 
This is a great way to check the string you entered really does represent the string you wanted.


In [3]:
# Putting lines in a vector
lines <- c(line1, line2, line3)

# Print lines
print(lines)

# Use writeLines() on lines
writeLines(lines)

# Write lines with a space separator
writeLines(lines, sep = " ")

# Use writeLines() on the string "hello\n\U1F30D"
writeLines("hello\n\U1F30D")

[1] "The table was a large one, but the three were all crowded together at one corner of it:"                           
[2] "\"No room! No room!\" they cried out when they saw Alice coming."                                                  
[3] "\"There's plenty of room!\" said Alice indignantly, and she sat down in a large arm-chair at one end of the table."
The table was a large one, but the three were all crowded together at one corner of it:
"No room! No room!" they cried out when they saw Alice coming.
"There's plenty of room!" said Alice indignantly, and she sat down in a large arm-chair at one end of the table.
The table was a large one, but the three were all crowded together at one corner of it: "No room! No room!" they cried out when they saw Alice coming. "There's plenty of room!" said Alice indignantly, and she sat down in a large arm-chair at one end of the table. hello
<U+0001F30D>


### Escape sequences
You might have been surprised at the output from the last part of the last exercise. 
How did you get two lines from one string? The key is the \ .
A sequence in a string that starts with a \ is called an escape sequence and allows us to include special 
characters in our strings. You saw one escape sequence in the first exercise: \" is used to denote a double quote.
In "hello\n\U1F30D" there are two escape sequences: \n gives a newline, and \U followed by up to 8 hex digits 
sequence denotes a particular Unicode character.
Unicode is a standard for representing characters that might not be on your keyboard. Each available character has 
a Unicode code point: a number that uniquely identifies it. These code points are generally written in hex notation, 
that is, using base 16 and the digits 0-9 and A-F. You can find the code point for a particular character by looking 
up a code chart. If you only need four digits for the codepoint, an alternative escape sequence is \u.
When R comes across a \ it assumes you are starting an escape, so if you actually need a backslash in your string 
you'll need the sequence \\.

In [4]:
# Edit the string inside writeLines() so that it correctly displays (all on one line)
# Should display: To have a \ you need \\
writeLines("To have a\\ you need \\\\")

# Should display: 
# This is a really 
# really really 
# long string
writeLines("This is a really\nreally really\nlong string")


To have a\ you need \\
This is a really
really really
long string


### Using format() with numbers
The behavior of format() can be pretty confusing, so you'll spend most of this exercise exploring how it works.

Recall from the video, the scientific argument to format() controls whether the numbers are displayed in fixed (scientific = FALSE) or scientific (scientific = TRUE) format.

When the representation is scientific, the digits argument is the number of digits before the exponent. When the representation is fixed, digits controls the significant digits used for the smallest (in magnitude) number. Each other number will be formatted to match the number of decimal places in the smallest number. This means the number of decimal places you get in your output depends on all the values you are formatting!

For example, if the smallest number is 0.0011, and digits = 1, then 0.0011 requires 3 places after the decimal to represent it to 1 significant digit, 0.001. Every other number will be formatted to 3 places after the decimal point.

So, how many decimal places will you get if 1.0011 is the smallest number? You'll find out in this exercise.

In [5]:
# Some vectors of numbers
percent_change  <- c(4, -1.91, 3.00, -5.002)
income <-  c(72.19, 1030.18, 10291.93, 1189192.18)
p_values <- c(0.12, 0.98, 0.0000191, 0.00000000002)

# Format c(0.0011, 0.011, 1) with digits = 1
format(c(0.0011, 0.011, 1), digits = 1)

# Format c(1.0011, 2.011, 1) with digits = 1
format(c(1.0011, 2.011, 1), digits = 1)

# Format percent_change to one place after the decimal point
format(percent_change, digits = 2)

# Format income to whole numbers (i.e. no places after the decimal point).
format(income, digits = 1, scientific = FALSE)

# Format p_values in fixed format
format(p_values, digits = 1, scientific = FALSE)

[1] "0.001" "0.011" "1.000"

[1] "1" "2" "1"

[1] " 4.0" "-1.9" " 3.0" "-5.0"

[1] "     72" "   1030" "  10292" "1189192"

[1] "0.12000000000" "0.98000000000" "0.00001910000" "0.00000000002"

### Controlling other aspects of the string
Not only does format() control the way the number is represented, it also controls some of the properties of the resulting string that affect its display.

For example, by default format() will pad the start of the strings with spaces so that the decimal points line up, which is really useful if you are presenting the numbers in a vertical column. However, if you are putting the number in the middle of a sentence, you might not want these extra spaces. You can set trim = TRUE to remove them.

When numbers are long it can be helpful to "prettify" them, for example instead of 1000000000 display 1,000,000,000. In this case a , is added every 3 digits. This can be controlled by the big.interval and big.mark arguments, e.g. format(1000000000, big.mark = ",", big.interval = 3, scientific = FALSE). These arguments are actually passed on to prettyNum() so head there for any further details.

In [6]:
income = c(10.67, 89.56, 2366.56, 24450.78)

formatted_income <- format(income, digits = 2)

# Print formatted_income
print(formatted_income)

# Call writeLines() on the formatted income
writeLines(formatted_income)

# Define trimmed_income
trimmed_income <- format(income, digits = 2, trim = TRUE)

# Call writeLines() on the trimmed_income
writeLines(trimmed_income)

# Define pretty_income
pretty_income = format(income, digits = 2, big.mark = ",")

# Call writeLines() on the pretty_income
writeLines(pretty_income)

[1] "   11" "   90" " 2367" "24451"
   11
   90
 2367
24451
11
90
2367
24451
    11
    90
 2,367
24,451


### formatC()
The function formatC() provides an alternative way to format numbers based on C style syntax.

Rather than a scientific argument, formatC() has a format argument that takes a code representing the required format. The most useful are:

"f" for fixed,
"e" for scientific, and
"g" for fixed unless scientific saves space
When using scientific format, the digits argument behaves like it does in format(); it specifies the number of significant digits. However, unlike format(), when using fixed format, digits is the number of digits after the decimal point. This is more predictable than format(), because the number of places after the decimal is fixed regardless of the values being formatted.

formatC() also formats numbers individually, which means you always get the same output regardless of other numbers in the vector.

The flag argument allows you to provide some modifiers that, for example, force the display of the sign (flag = "+"), left align numbers (flag = "-") and pad numbers with leading zeros (flag = "0"). You'll see an example in this exercise.

In [8]:
# From the format() exercise
x <- c(0.0011, 0.011, 1)
y <- c(1.0011, 2.011, 1)

# formatC() on x with format = "f", digits = 1
formatC(x, format = "f", digits = 1)

# formatC() on y with format = "f", digits = 1
formatC(y, format = "f", digits = 1)
# Format percent_change to one place after the decimal point
formatC(percent_change, format = 'f', digits = 1)

# percent_change with flag = "+"
formatC(percent_change, format = 'f', digits = 1, flag = '+')


# Format p_values using format = "g" and digits = 2
formatC(p_values, digits = 2, format = 'g')

[1] "0.0" "0.0" "1.0"

[1] "1.0" "2.0" "1.0"

[1] "4.0"  "-1.9" "3.0"  "-5.0"

[1] "+4.0" "-1.9" "+3.0" "-5.0"

[1] "0.12"    "0.98"    "1.9e-05" "2e-11"

### Annotation of numbers
To get a handle on using paste(), you are going to annotate some of your formatted number strings.

The key points to remember are:

The vectors you pass to paste() are pasted together element by element, using the sep argument to combine them.
If the vectors passed to paste() aren't the same length, the shorter vectors are recycled up to the length of the longest one.
Only use collapse if you want a single string as output. collapse specifies the string to place between different elements.

We've put the formatted vectors pretty_income and pretty_percent in your workspace along with years.

In [11]:
# Define pretty_income
pretty_income = format(income, digits = 2, big.mark = ",")
# define pretty_percent
pretty_percent = formatC(percent_change, format = 'f', digits = 1, flag = '+')
# define years 
years = c(2010,2011, 2012, 2013)

# Add $ to pretty_income
paste("$", pretty_income, sep = "")
 
# Add % to pretty_percent
paste(pretty_percent, "%", sep = "")

# Create vector with elements like 2010: +4.0%`
year_percent <- paste(years, ": ", pretty_percent, "%", sep = "")

# Collapse all years into single string
paste(year_percent, collapse = ", ")

[1] "$    11" "$    90" "$ 2,367" "$24,451"

[1] "+4.0%" "-1.9%" "+3.0%" "-5.0%"

[1] "2010: +4.0%, 2011: -1.9%, 2012: +3.0%, 2013: -5.0%"

### A very simple table
Combining format() and paste() is one way to display very simple tables. Remember, since format() looks at all the values in a vector before formatting, it uses a consistent format and will, by default, align on the decimal point. This is usually the behavior you want for a column of numbers in table.

format() can also take character vectors as input. In this case, you can use the justify argument, specific to character input, to justify the text to the left, right, or center.

You are going to put together the following table:

Year 0   $       72

Year 1   $    1,030

Year 2   $   10,292

Project Lifetime   $1,189,192

You'll start by formatting the columns to prepare to put them in a table, then you'll use paste() to put together each row. Then, you can use writeLines() to display each row on a new line.

In [12]:
# Define the names vector
income_names <- c("Year 0", "Year 1", "Year 2", "Project Lifetime")

# Create pretty_income
pretty_income <- format(income, digits = 2, big.mark = ",")

# Create dollar_income
dollar_income <- paste("$", pretty_income, sep = "")

# Create formatted_names
formatted_names <- format(income_names, justify = "right")

# Create rows
rows <- paste(formatted_names, dollar_income, sep = "   ")

# Write rows
writeLines(rows)

          Year 0   $    11
          Year 1   $    90
          Year 2   $ 2,367
Project Lifetime   $24,451


### Let's order pizza!
As a final exercise in using paste() and to celebrate getting to the end of the first chapter, let's order some pizza.

We've got a list of possible pizza toppings in toppings.

You are going to randomly select three toppings, and then put them together using paste() into an order for pizza, that should result in a string like,

"I want to order a pizza with mushrooms, spinach, and pineapple."

In [14]:
# Randomly sample 3 toppings
toppings = c("anchovies","artichoke", "bacon", "breakfast bacon" , "Canadian bacon" , "cheese", "chicken", "chili peppers")   
my_toppings <- sample(toppings, size = 3)

# Print my_toppings
my_toppings

# Paste "and " to last element: my_toppings_and
my_toppings_and <- paste(c("", "", "and "), my_toppings, sep = "")
# my_toppings_and <- c(my_toppings[1], my_toppings[2], paste("and ", my_toppings[3]))

# Collapse with comma space: these_toppings
these_toppings <- paste(my_toppings_and, collapse = ", ")

# Add rest of sentence: my_order
my_order <- paste("I want to order a pizza with ", these_toppings, ".", sep = "")

# Order pizza with writeLines()
writeLines(my_order)

[1] "chicken"       "bacon"         "chili peppers"

I want to order a pizza with chicken, bacon, and chili peppers.
